File: src/lib/calculations/cashflow.ts
/**
 * CRE Cash Flow Calculation Engine
 *
 * Calculates property-level cash flow for commercial real estate properties
 * Supports multiple lease types: NNN, Modified Gross, Gross, Ground Lease
 *
 * Based on ARGUS-level retail analysis methodology
 */

export interface LeaseData {
  lease_id: number;
  space_id: number;
  tenant_id: number;
  tenant_name: string;
  lease_type: 'NNN' | 'Modified Gross' | 'Gross' | 'Ground Lease' | 'Absolute NNN';
  lease_status: string;
  lease_commencement_date: Date;
  lease_expiration_date: Date;
  leased_sf: number;

  // Base rent schedule
  base_rent: {
    period_start_date: Date;
    period_end_date: Date;
    base_rent_annual: number;
    base_rent_monthly: number;
    base_rent_psf_annual: number;
  }[];

  // Escalations
  escalation?: {
    escalation_type: 'Fixed Percentage' | 'CPI' | 'Fixed Dollar' | 'Stepped';
    escalation_pct?: number;
    escalation_frequency?: string;
    cpi_floor_pct?: number;
    cpi_cap_pct?: number;
  };

  // Percentage rent (retail)
  percentage_rent?: {
    breakpoint_amount: number;
    percentage_rate: number;
    prior_year_sales?: number;
  };

  // Expense recovery structure
  expense_recovery: {
    recovery_structure: string;
    property_tax_recovery_pct: number;
    insurance_recovery_pct: number;
    cam_recovery_pct: number;
    utilities_recovery_pct?: number;
    expense_cap_psf?: number;
    expense_cap_escalation_pct?: number;
  };
}

export interface PropertyData {
  cre_property_id: number;
  property_name: string;
  rentable_sf: number;
  acquisition_price: number;
  leases: LeaseData[];
}

export interface OperatingExpenses {
  period_date: Date;
  property_taxes: number;
  insurance: number;
  cam_expenses: number;
  utilities: number;
  management_fee_pct: number; // As % of EGI
  repairs_maintenance: number;
  other_expenses: number;
}

export interface CapitalItems {
  period_date: Date;
  capital_reserves: number;
  tenant_improvements: number;
  leasing_commissions: number;
}

export interface CashFlowPeriod {
  period_id: number;
  period_date: Date;

  // Revenue components
  base_rent_revenue: number;
  percentage_rent_revenue: number;
  expense_recovery_revenue: number;
  other_income: number;
  gross_revenue: number;

  // Vacancy & credit loss
  vacancy_loss: number;
  credit_loss: number;
  effective_gross_income: number;

  // Operating expenses
  property_taxes: number;
  insurance: number;
  cam_expenses: number;
  utilities: number;
  management_fee: number;
  repairs_maintenance: number;
  other_expenses: number;
  total_operating_expenses: number;

  // NOI
  net_operating_income: number;

  // Capital items
  capital_reserves: number;
  tenant_improvements: number;
  leasing_commissions: number;
  total_capital: number;

  // Cash flow
  cash_flow_before_debt: number;
  debt_service: number;
  net_cash_flow: number;
}

/**
 * Calculate base rent for a specific period
 */
export function calculateBaseRent(
  lease: LeaseData,
  periodDate: Date
): number {
  // Find the applicable rent schedule for this period
  const applicableRent = lease.base_rent.find(rent => {
    const startDate = new Date(rent.period_start_date);
    const endDate = new Date(rent.period_end_date);
    return periodDate >= startDate && periodDate <= endDate;
  });

  if (!applicableRent) {
    // Lease may not have commenced or has expired
    return 0;
  }

  // Return monthly rent (annual / 12)
  return applicableRent.base_rent_annual / 12;
}

/**
 * Calculate percentage rent for retail tenants
 */
export function calculatePercentageRent(
  lease: LeaseData,
  periodDate: Date,
  projectedAnnualSales?: number
): number {
  if (!lease.percentage_rent) {
    return 0;
  }

  const { breakpoint_amount, percentage_rate, prior_year_sales } = lease.percentage_rent;

  // Use projected sales or prior year sales
  const annualSales = projectedAnnualSales || prior_year_sales || 0;

  if (annualSales <= breakpoint_amount) {
    return 0; // Sales below breakpoint
  }

  // Calculate overage rent (monthly)
  const overageAmount = annualSales - breakpoint_amount;
  const percentageRentAnnual = overageAmount * (percentage_rate / 100);

  return percentageRentAnnual / 12; // Monthly
}

/**
 * Calculate expense recoveries from tenant based on lease structure
 */
export function calculateExpenseRecovery(
  lease: LeaseData,
  opex: OperatingExpenses,
  totalRentableSF: number
): number {
  const { expense_recovery } = lease;
  const tenantProRataShare = lease.leased_sf / totalRentableSF;

  // Calculate tenant's share of each expense category
  const taxRecovery = opex.property_taxes * tenantProRataShare * (expense_recovery.property_tax_recovery_pct / 100);
  const insuranceRecovery = opex.insurance * tenantProRataShare * (expense_recovery.insurance_recovery_pct / 100);
  const camRecovery = opex.cam_expenses * tenantProRataShare * (expense_recovery.cam_recovery_pct / 100);
  const utilityRecovery = opex.utilities * tenantProRataShare * ((expense_recovery.utilities_recovery_pct || 0) / 100);

  let totalRecovery = taxRecovery + insuranceRecovery + camRecovery + utilityRecovery;

  // Apply CAM cap if specified (Modified Gross leases)
  if (expense_recovery.expense_cap_psf) {
    const capAmount = expense_recovery.expense_cap_psf * lease.leased_sf / 12; // Monthly cap
    totalRecovery = Math.min(totalRecovery, capAmount);
  }

  return totalRecovery;
}

/**
 * Calculate property-level cash flow for a single period
 */
export function calculatePeriodCashFlow(
  property: PropertyData,
  periodDate: Date,
  periodNumber: number,
  opex: OperatingExpenses,
  capitalItems: CapitalItems,
  debtService: number = 0,
  vacancyPct: number = 0.05,
  creditLossPct: number = 0.02
): CashFlowPeriod {
  // Initialize revenue components
  let baseRentRevenue = 0;
  let percentageRentRevenue = 0;
  let expenseRecoveryRevenue = 0;

  // Calculate revenue from all active leases
  for (const lease of property.leases) {
    const leaseStartDate = new Date(lease.lease_commencement_date);
    const leaseEndDate = new Date(lease.lease_expiration_date);

    // Skip if lease not active in this period
    if (periodDate < leaseStartDate || periodDate > leaseEndDate) {
      continue;
    }

    // Base rent
    baseRentRevenue += calculateBaseRent(lease, periodDate);

    // Percentage rent (if applicable)
    percentageRentRevenue += calculatePercentageRent(lease, periodDate);

    // Expense recoveries
    expenseRecoveryRevenue += calculateExpenseRecovery(lease, opex, property.rentable_sf);
  }

  const otherIncome = 0; // Parking, signage, etc. - could be added
  const grossRevenue = baseRentRevenue + percentageRentRevenue + expenseRecoveryRevenue + otherIncome;

  // Vacancy and credit loss
  const vacancyLoss = grossRevenue * vacancyPct;
  const creditLoss = grossRevenue * creditLossPct;
  const effectiveGrossIncome = grossRevenue - vacancyLoss - creditLoss;

  // Operating expenses
  const managementFee = effectiveGrossIncome * (opex.management_fee_pct / 100);
  const totalOperatingExpenses =
    opex.property_taxes +
    opex.insurance +
    opex.cam_expenses +
    opex.utilities +
    managementFee +
    opex.repairs_maintenance +
    opex.other_expenses;

  // Net Operating Income
  const netOperatingIncome = effectiveGrossIncome - totalOperatingExpenses;

  // Capital items
  const totalCapital =
    capitalItems.capital_reserves +
    capitalItems.tenant_improvements +
    capitalItems.leasing_commissions;

  // Cash flow
  const cashFlowBeforeDebt = netOperatingIncome - totalCapital;
  const netCashFlow = cashFlowBeforeDebt - debtService;

  return {
    period_id: periodNumber,
    period_date: periodDate,

    // Revenue
    base_rent_revenue: baseRentRevenue,
    percentage_rent_revenue: percentageRentRevenue,
    expense_recovery_revenue: expenseRecoveryRevenue,
    other_income: otherIncome,
    gross_revenue: grossRevenue,

    // Vacancy & credit loss
    vacancy_loss: vacancyLoss,
    credit_loss: creditLoss,
    effective_gross_income: effectiveGrossIncome,

    // Operating expenses
    property_taxes: opex.property_taxes,
    insurance: opex.insurance,
    cam_expenses: opex.cam_expenses,
    utilities: opex.utilities,
    management_fee: managementFee,
    repairs_maintenance: opex.repairs_maintenance,
    other_expenses: opex.other_expenses,
    total_operating_expenses: totalOperatingExpenses,

    // NOI
    net_operating_income: netOperatingIncome,

    // Capital
    capital_reserves: capitalItems.capital_reserves,
    tenant_improvements: capitalItems.tenant_improvements,
    leasing_commissions: capitalItems.leasing_commissions,
    total_capital: totalCapital,

    // Cash flow
    cash_flow_before_debt: cashFlowBeforeDebt,
    debt_service: debtService,
    net_cash_flow: netCashFlow,
  };
}

/**
 * Calculate cash flow for entire hold period (all periods)
 */
export function calculateMultiPeriodCashFlow(
  property: PropertyData,
  startDate: Date,
  numPeriods: number,
  periodType: 'monthly' | 'annual',
  opexSchedule: OperatingExpenses[],
  capitalSchedule: CapitalItems[],
  debtServiceAnnual: number = 0,
  vacancyPct: number = 0.05,
  creditLossPct: number = 0.02
): CashFlowPeriod[] {
  const cashFlows: CashFlowPeriod[] = [];
  const monthsIncrement = periodType === 'monthly' ? 1 : 12;
  const debtServicePeriod = periodType === 'monthly' ? debtServiceAnnual / 12 : debtServiceAnnual;

  for (let i = 0; i < numPeriods; i++) {
    const periodDate = new Date(startDate);
    periodDate.setMonth(periodDate.getMonth() + (i * monthsIncrement));

    // Find applicable opex and capital for this period
    const opex = opexSchedule[i] || opexSchedule[0]; // Use first period if not specified
    const capital = capitalSchedule[i] || { period_date: periodDate, capital_reserves: 0, tenant_improvements: 0, leasing_commissions: 0 };

    const cashFlow = calculatePeriodCashFlow(
      property,
      periodDate,
      i + 1,
      opex,
      capital,
      debtServicePeriod,
      vacancyPct,
      creditLossPct
    );

    cashFlows.push(cashFlow);
  }

  return cashFlows;
}

/**
 * Calculate Year 1 NOI (used for cap rate valuation)
 */
export function calculateYear1NOI(cashFlows: CashFlowPeriod[]): number {
  if (cashFlows.length === 0) return 0;

  // Sum first 12 months of NOI (or all periods if less than 12)
  const year1Periods = cashFlows.slice(0, Math.min(12, cashFlows.length));
  return year1Periods.reduce((sum, cf) => sum + cf.net_operating_income, 0);
}

/**
 * Calculate stabilized NOI (typically year 2-5 average, or last year)
 */
export function calculateStabilizedNOI(cashFlows: CashFlowPeriod[], holdPeriodYears: number): number {
  if (cashFlows.length === 0) return 0;

  // Use last year as stabilized (terminal year)
  const lastYearStart = Math.max(0, cashFlows.length - 12);
  const lastYearPeriods = cashFlows.slice(lastYearStart);

  return lastYearPeriods.reduce((sum, cf) => sum + cf.net_operating_income, 0);
}

File: backend/financial_engine/core/cashflow.py
"""
Cash Flow Projection Engine

Calculates multi-period cash flow projections for commercial properties using
pandas DataFrames for vectorized operations (5-10x faster than TypeScript).

This replaces src/lib/calculations/cashflow.ts (381 lines).
"""

import numpy as np
import pandas as pd
from datetime import date, timedelta
from typing import List, Optional, Literal
from loguru import logger

from financial_engine.models import (
    PropertyData,
    LeaseData,
    OperatingExpenses,
    CapitalItems,
    CashFlowPeriod,
)
from financial_engine.config import get_settings


class CashFlowEngine:
    """
    Multi-period cash flow projection engine for CRE properties.

    Uses pandas DataFrames for vectorized calculations - much faster than
    iterating through periods in TypeScript.

    Example:
        >>> engine = CashFlowEngine()
        >>> cash_flows = engine.calculate_multi_period_cashflow(
        ...     property=property_data,
        ...     start_date=date(2025, 1, 1),
        ...     num_periods=120,
        ...     period_type="monthly",
        ...     opex_schedule=monthly_opex,
        ...     capital_schedule=monthly_capital,
        ... )
        >>> df = pd.DataFrame([cf.dict() for cf in cash_flows])
        >>> print(f"Total NOI: ${df['net_operating_income'].sum():,.0f}")
    """

    def __init__(self):
        self.settings = get_settings()

    def calculate_multi_period_cashflow(
        self,
        property: PropertyData,
        start_date: date,
        num_periods: int,
        opex_schedule: List[OperatingExpenses],
        capital_schedule: List[CapitalItems],
        period_type: Literal["monthly", "annual"] = "monthly",
        annual_debt_service: float = 0,
        vacancy_pct: Optional[float] = None,
        credit_loss_pct: Optional[float] = None,
    ) -> List[CashFlowPeriod]:
        """
        Calculate multi-period cash flow projections.

        This is the main entry point for cash flow calculations.

        Args:
            property: Property data with lease schedules
            start_date: Start date for projections
            num_periods: Number of periods to project
            period_type: "monthly" or "annual"
            opex_schedule: Operating expenses by period
            capital_schedule: Capital items by period
            annual_debt_service: Annual debt service (if levered)
            vacancy_pct: Vacancy percentage (defaults to config)
            credit_loss_pct: Credit loss percentage (defaults to config)

        Returns:
            List of CashFlowPeriod objects (one per period)

        Example:
            >>> cash_flows = engine.calculate_multi_period_cashflow(
            ...     property=property_data,
            ...     start_date=date(2025, 1, 1),
            ...     num_periods=120,
            ...     period_type="monthly",
            ...     opex_schedule=[monthly_opex] * 120,
            ...     capital_schedule=[monthly_capital] * 120,
            ...     annual_debt_service=483_000,
            ...     vacancy_pct=0.05,
            ...     credit_loss_pct=0.02,
            ... )
        """
        vacancy_pct = vacancy_pct or self.settings.default_vacancy_pct
        credit_loss_pct = credit_loss_pct or self.settings.default_credit_loss_pct

        logger.info(
            f"Calculating {num_periods} {period_type} cash flow periods "
            f"starting {start_date.isoformat()}"
        )

        # Generate period dates
        period_dates = self._generate_period_dates(start_date, num_periods, period_type)

        # Calculate base rent by period for all leases (vectorized)
        lease_revenue_df = self._calculate_lease_revenue(
            property.leases, period_dates, period_type
        )

        # Calculate expense recoveries by period
        recovery_revenue_df = self._calculate_expense_recoveries(
            property.leases,
            opex_schedule,
            period_dates,
        )

        # Combine all revenue streams
        total_revenue = lease_revenue_df.sum(axis=1) + recovery_revenue_df.sum(axis=1)

        # Apply vacancy and credit loss
        vacancy_loss = total_revenue * vacancy_pct
        credit_loss = (total_revenue - vacancy_loss) * credit_loss_pct
        effective_gross_income = total_revenue - vacancy_loss - credit_loss

        # Calculate operating expenses
        opex_df = self._calculate_operating_expenses(
            opex_schedule, effective_gross_income, period_dates
        )
        total_opex = opex_df.sum(axis=1)

        # Net Operating Income
        noi = effective_gross_income - total_opex

        # Capital expenses
        capital_df = self._calculate_capital_expenses(capital_schedule)
        total_capital = capital_df.sum(axis=1)

        # Debt service (monthly if monthly periods, annual if annual)
        if period_type == "monthly":
            debt_service = annual_debt_service / 12
        else:
            debt_service = annual_debt_service

        # Net Cash Flow
        net_cash_flow = noi - total_capital - debt_service

        # Build result objects
        cash_flows = []
        for i in range(num_periods):
            period = CashFlowPeriod(
                period_number=i + 1,
                period_date=period_dates[i],
                # Revenue
                base_rent_revenue=float(lease_revenue_df.iloc[i].sum()),
                expense_recovery_revenue=float(recovery_revenue_df.iloc[i].sum()),
                other_income=0.0,  # TODO: Add parking, storage, etc.
                gross_potential_income=float(total_revenue.iloc[i]),
                vacancy_loss=float(vacancy_loss.iloc[i]),
                credit_loss=float(credit_loss.iloc[i]),
                effective_gross_income=float(effective_gross_income.iloc[i]),
                # Operating Expenses
                property_taxes=float(opex_df["property_taxes"].iloc[i]),
                insurance=float(opex_df["insurance"].iloc[i]),
                utilities=float(opex_df["utilities"].iloc[i]),
                repairs_maintenance=float(opex_df["repairs_maintenance"].iloc[i]),
                management_fee=float(opex_df["management_fee"].iloc[i]),
                other_operating_expenses=float(
                    opex_df["cam_expenses"].iloc[i] + opex_df["other_expenses"].iloc[i]
                ),
                total_operating_expenses=float(total_opex.iloc[i]),
                net_operating_income=float(noi.iloc[i]),
                # Capital & Debt
                capital_expenditures=float(total_capital.iloc[i]),
                debt_service=debt_service,
                net_cash_flow=float(net_cash_flow.iloc[i]),
            )
            cash_flows.append(period)

        logger.info(
            f"Calculated {len(cash_flows)} periods: "
            f"Total NOI=${sum(cf.net_operating_income for cf in cash_flows):,.0f}, "
            f"Total NCF=${sum(cf.net_cash_flow for cf in cash_flows):,.0f}"
        )

        return cash_flows

    def _generate_period_dates(
        self, start_date: date, num_periods: int, period_type: str
    ) -> pd.DatetimeIndex:
        """Generate period dates based on start date and period type."""
        if period_type == "monthly":
            dates = pd.date_range(start=start_date, periods=num_periods, freq="MS")
        else:  # annual
            dates = pd.date_range(start=start_date, periods=num_periods, freq="YS")

        return dates

    def _calculate_lease_revenue(
        self, leases: List[LeaseData], period_dates: pd.DatetimeIndex, period_type: str
    ) -> pd.DataFrame:
        """
        Calculate base rent revenue for all leases across all periods.

        Returns DataFrame with columns = lease_ids, rows = periods.
        Uses vectorized operations for performance.
        """
        revenue_data = {}

        for lease in leases:
            lease_revenue = []

            for period_date in period_dates:
                # Find applicable base rent schedule for this period
                applicable_rent = None
                for rent_schedule in lease.base_rent:
                    if (
                        rent_schedule.period_start_date <= period_date.date()
                        <= rent_schedule.period_end_date
                    ):
                        applicable_rent = rent_schedule
                        break

                if applicable_rent:
                    if period_type == "monthly":
                        revenue = applicable_rent.base_rent_monthly
                    else:
                        revenue = applicable_rent.base_rent_annual
                else:
                    # Lease not active in this period
                    revenue = 0.0

                lease_revenue.append(revenue)

            revenue_data[f"lease_{lease.lease_id}"] = lease_revenue

        df = pd.DataFrame(revenue_data, index=period_dates)
        return df

    def _calculate_expense_recoveries(
        self,
        leases: List[LeaseData],
        opex_schedule: List[OperatingExpenses],
        period_dates: pd.DatetimeIndex,
    ) -> pd.DataFrame:
        """
        Calculate expense recovery revenue from tenants.

        For NNN/Modified Gross leases, tenants reimburse operating expenses.
        """
        recovery_data = {}

        for lease in leases:
            lease_recoveries = []

            for i, period_date in enumerate(period_dates):
                # Check if lease is active in this period
                if not (
                    lease.lease_commencement_date
                    <= period_date.date()
                    <= lease.lease_expiration_date
                ):
                    lease_recoveries.append(0.0)
                    continue

                opex = opex_schedule[i]
                recovery = lease.expense_recovery

                # Calculate recoverable expenses based on recovery structure
                if recovery.recovery_structure == "Triple Net (NNN)":
                    # Tenant pays all operating expenses proportionally
                    recoverable = (
                        opex.property_taxes * (recovery.property_tax_recovery_pct / 100)
                        + opex.insurance * (recovery.insurance_recovery_pct / 100)
                        + opex.cam_expenses * (recovery.cam_recovery_pct / 100)
                    )
                elif recovery.recovery_structure == "Modified Gross":
                    # Tenant pays CAM, insurance, and taxes (but not management/repairs)
                    recoverable = (
                        opex.property_taxes * (recovery.property_tax_recovery_pct / 100)
                        + opex.insurance * (recovery.insurance_recovery_pct / 100)
                        + opex.cam_expenses * (recovery.cam_recovery_pct / 100)
                    )
                else:  # Gross lease
                    recoverable = 0.0

                # Apply expense cap if defined
                if recovery.expense_cap_psf > 0:
                    max_recovery = recovery.expense_cap_psf * lease.leased_sf / 12
                    recoverable = min(recoverable, max_recovery)

                lease_recoveries.append(recoverable)

            recovery_data[f"recovery_{lease.lease_id}"] = lease_recoveries

        df = pd.DataFrame(recovery_data, index=period_dates)
        return df

    def _calculate_operating_expenses(
        self,
        opex_schedule: List[OperatingExpenses],
        effective_gross_income: pd.Series,
        period_dates: pd.DatetimeIndex,
    ) -> pd.DataFrame:
        """
        Calculate operating expenses by period.

        Management fee is calculated as % of EGI (varies by period).
        Other expenses are fixed by period.
        """
        opex_data = {
            "property_taxes": [],
            "insurance": [],
            "cam_expenses": [],
            "utilities": [],
            "repairs_maintenance": [],
            "management_fee": [],
            "other_expenses": [],
        }

        for i, opex in enumerate(opex_schedule):
            # Management fee is % of EGI
            management_fee = effective_gross_income.iloc[i] * (
                opex.management_fee_pct / 100
            )

            opex_data["property_taxes"].append(opex.property_taxes)
            opex_data["insurance"].append(opex.insurance)
            opex_data["cam_expenses"].append(opex.cam_expenses)
            opex_data["utilities"].append(opex.utilities)
            opex_data["repairs_maintenance"].append(opex.repairs_maintenance)
            opex_data["management_fee"].append(management_fee)
            opex_data["other_expenses"].append(opex.other_expenses)

        df = pd.DataFrame(opex_data, index=period_dates)
        return df

    def _calculate_capital_expenses(
        self, capital_schedule: List[CapitalItems]
    ) -> pd.DataFrame:
        """
        Calculate capital expenses by period.

        Includes capital reserves, TIs, leasing commissions.
        """
        capital_data = {
            "capital_reserves": [],
            "tenant_improvements": [],
            "leasing_commissions": [],
        }

        for capital in capital_schedule:
            capital_data["capital_reserves"].append(capital.capital_reserves)
            capital_data["tenant_improvements"].append(capital.tenant_improvements)
            capital_data["leasing_commissions"].append(capital.leasing_commissions)

        df = pd.DataFrame(capital_data)
        return df

    def calculate_annual_summary(
        self, cash_flows: List[CashFlowPeriod]
    ) -> pd.DataFrame:
        """
        Aggregate monthly cash flows into annual summary.

        Useful for reporting and presentation.

        Args:
            cash_flows: List of monthly CashFlowPeriod objects

        Returns:
            DataFrame with annual aggregations

        Example:
            >>> annual_df = engine.calculate_annual_summary(monthly_cash_flows)
            >>> print(annual_df[["year", "noi", "ncf"]])
               year         noi          ncf
            0  2025  1,250,000    890,000
            1  2026  1,287,500    925,000
        """
        # Convert to DataFrame
        df = pd.DataFrame([cf.dict() for cf in cash_flows])
        df["period_date"] = pd.to_datetime(df["period_date"])
        df["year"] = df["period_date"].dt.year

        # Aggregate by year
        annual_df = df.groupby("year").agg(
            {
                "gross_potential_income": "sum",
                "vacancy_loss": "sum",
                "credit_loss": "sum",
                "effective_gross_income": "sum",
                "total_operating_expenses": "sum",
                "net_operating_income": "sum",
                "capital_expenditures": "sum",
                "debt_service": "sum",
                "net_cash_flow": "sum",
            }
        )

        annual_df.reset_index(inplace=True)

        logger.debug(f"Aggregated {len(cash_flows)} periods into {len(annual_df)} years")

        return annual_df

    def export_to_excel(
        self, cash_flows: List[CashFlowPeriod], output_path: str
    ) -> None:
        """
        Export cash flow projections to Excel file.

        Args:
            cash_flows: List of CashFlowPeriod objects
            output_path: Path to output Excel file

        Example:
            >>> engine.export_to_excel(cash_flows, "cash_flow_projections.xlsx")
        """
        df = pd.DataFrame([cf.dict() for cf in cash_flows])

        # Format dates
        df["period_date"] = pd.to_datetime(df["period_date"]).dt.strftime("%Y-%m-%d")

        # Write to Excel with formatting
        with pd.ExcelWriter(output_path, engine="openpyxl") as writer:
            df.to_excel(writer, sheet_name="Cash Flow", index=False)

            # Add annual summary sheet
            annual_df = self.calculate_annual_summary(cash_flows)
            annual_df.to_excel(writer, sheet_name="Annual Summary", index=False)

        logger.info(f"Exported cash flow projections to {output_path}")
